packages/next/src/build/index.ts

```diff
@@ -933,6 +935,89 @@ export default async function build(
         })
       )
 
+      let turboTasks: unknown
+
+      if (turbotraceContext) {
+        let binding = (await loadBindings()) as any
+        if (
+          !binding?.isWasm &&
+          typeof binding.turbo.startTrace === 'function'
+        ) {
+          let turbotraceOutputPath: string | undefined
+          let turbotraceFiles: string[] | undefined
+          turboTasks = binding.turbo.createTurboTasks(
+            config.experimental.turbotrace?.memoryLimit ??
+              TURBO_TRACE_DEFAULT_MEMORY_LIMIT
+          )
+
+          const { entriesTrace, chunksTrace } = turbotraceContext
+          if (entriesTrace) {
+            const {
+              appDir: turbotraceContextAppDir,
+              depModArray,
+              entryNameMap,
+              outputPath,
+              action,
+            } = entriesTrace
+            const depModSet = new Set(depModArray)
+            const filesTracedInEntries: string[] =
+              await binding.turbo.startTrace(action, turboTasks)
+
+            const { contextDirectory, input: entriesToTrace } = action
+
+            // only trace the assets under the appDir
+            // exclude files from node_modules, entries and processed by webpack
+            const filesTracedFromEntries = filesTracedInEntries
+              .map((f) => path.join(contextDirectory, f))
+              .filter(
+                (f) =>
+                  !f.includes('/node_modules/') &&
+                  f.startsWith(turbotraceContextAppDir) &&
+                  !entriesToTrace.includes(f) &&
+                  !depModSet.has(f)
+              )
+            if (filesTracedFromEntries.length) {
+              // The turbo trace doesn't provide the traced file type and reason at present
+              // let's write the traced files into the first [entry].nft.json
+              const [[, entryName]] = Array.from(entryNameMap.entries()).filter(
+                ([k]) => k.startsWith(turbotraceContextAppDir)
```

Comment:
Nit pick: this can use keys only
```suggestion
              const [[, entryName]] = Array.from(entryNameMap.keys()).filter(
                (k) => k.startsWith(turbotraceContextAppDir)
```